### PR TITLE
Aliases with ContestedResourceVoteState

### PR DESCRIPTION
**Pull request by owl352**
# Issue
For new design we need owner of transaction in api responses
Also we need aliases for identity based on contestedResourceVoteState
# Things done
* package.json again includes my forks for dapi
* Implemented new method in DAPI.js
* Implemented new method in utils.js, which allows check if alias is contested and make him normalized
* Updated alghoritms for identities searches in IdentitiesController.js for new logic with aliases
* Updated https://github.com/owl352/dapi-client and https://github.com/owl352/dapi-grpc
* Added owner field in request and model for transactions in Main, Transactions and Identity controllers
* Added trim for owner in transaction model
* Added aliases for transactions in transaction controller
* Improved main.spec.js
* Updated tests
* Updated `Readme.md`

***#302 MERGED HERE***

### CODE REVIEW

File: packages/api/package.json [hunk 1]
```diff
@@ -12,19 +12,19 @@
...
   },
   "dependencies": {
-    "@dashevo/dapi-client": "1.5.1",
     "@dashevo/dashd-rpc": "19.0.0",
     "@fastify/cors": "^8.3.0",
     "@scure/base": "^1.1.5",
     "bs58": "^6.0.0",
     "cbor": "^9.0.2",
-    "dash": "4.5.1",
+    "dash": "4.4.1",
```
Comment (pshenmic) [inline packages/api/package.json:20]: Do no revert

File: packages/api/src/DAPI.js [hunk 2]
```diff
@@ -23,6 +23,27 @@ class DAPI {
     const { epochsInfo } = await this.dapi.platform.getEpochsInfo(start, count, { ascending })
     return epochsInfo
   }
+
+  async getContestedState (contractId,
+    documentTypeName,
+    indexName = 'parentNameAndLabel',
```
Comment (pshenmic) [inline packages/api/src/DAPI.js:29]: should not default

File: packages/api/src/DAPI.js [hunk 3]
```diff
@@ -23,6 +23,27 @@ class DAPI {
...
+    resultType = 2,
```
Comment (pshenmic) [inline packages/api/src/DAPI.js:30]: should not default

File: packages/api/src/DAPI.js [hunk 4]
```diff
@@ -23,6 +23,27 @@ class DAPI {
     const { epochsInfo } = await this.dapi.platform.getEpochsInfo(start, count, { ascending })
     return epochsInfo
   }
+
+  async getContestedState (contractId,
```
Comment (pshenmic) [inline packages/api/src/DAPI.js:27]: add jsdoc

File: packages/api/src/DAPI.js [hunk 5]
```diff
@@ -23,6 +23,27 @@ class DAPI {
...
+    allowIncludeLockedAndAbstainingVoteTally,
+    count
+  ) {
+    const { contestedResourceContenders } = await this.dapi.platform.getContestedResourceVoteState(
+      Buffer.from(contractId, 'base64'),
+      documentTypeName,
+      indexName,
+      resultType,
+      indexValuesList,
+      startAtIdentifierInfo,
```
Comment (pshenmic) [inline packages/api/src/DAPI.js:41]: name params as in Dash SDK's getContestedResourceVoteState

File: packages/api/src/constants.js [hunk 6]
```diff
@@ -5,6 +5,7 @@ let genesisTime
 module.exports = {
   EPOCH_CHANGE_TIME: Number(process.env.EPOCH_CHANGE_TIME),
   TCP_CONNECT_TIMEOUT: Number(process.env.TCP_CONNECT_TIMEOUT),
+  DPNS_CONTRACT: process.env.DPNS_CONTRACT ?? '5mjGWa9mruHnLBht3ntbfgodcSoJxA1XIfYiv1PFMVU=',
```
Comment (pshenmic) [inline packages/api/src/constants.js:8]: I think we can just hardcode it, it does not change across networks

File: packages/api/src/controllers/IdentitiesController.js [hunk 7]
```diff
@@ -16,22 +19,55 @@ class IdentitiesController {
       return response.status(404).send({ message: 'not found' })
     }
 
+    const validatedAliases = await validateAliases(identity.aliases, identity.identifier, this.dapi)
```
Comment (pshenmic) [inline packages/api/src/controllers/IdentitiesController.js:22]: Let's move it in the DAO layer

File: packages/api/src/controllers/IdentitiesController.js [hunk 8]
```diff
@@ -16,22 +19,55 @@ class IdentitiesController {
...
+
     const balance = await this.dapi.getIdentityBalance(identifier)
 
-    response.send({ ...identity, balance })
+    response.send(Identity.fromObject({ ...identity, aliases: validatedAliases, balance }))
   }
 
   getIdentityByDPNS = async (request, response) => {
```
Comment (pshenmic) [inline packages/api/src/controllers/IdentitiesController.js:29]: getIdentityByDPNSName

File: packages/api/src/controllers/IdentitiesController.js [hunk 9]
```diff
@@ -16,22 +19,55 @@ class IdentitiesController {
...
     const { dpns } = request.query
 
-    const identity = await this.identitiesDAO.getIdentityByDPNS(dpns)
+    let preIdentity
+    let identity
+
+    if (!dpns.includes('.')) {
```
Comment (pshenmic) [inline packages/api/src/controllers/IdentitiesController.js:35]: Very complex logic. what does this do?

File: packages/api/src/controllers/MainController.js [hunk 10]
```diff
@@ -115,9 +117,9 @@ class MainController {
       const identity = await this.identitiesDAO.getIdentityByIdentifier(query)
 
       if (identity) {
-        const balance = await this.dapi.getIdentityBalance(identity.identifier)
-
-        return response.send({ identity: { ...identity, balance } })
+        // Sending without actual balance and aliases, because on frontend we were making
```
Comment (pshenmic) [inline packages/api/src/controllers/MainController.js:120]:
```suggestion
        // Sending without balance and aliases, not available on search, use 
```

File: packages/api/src/utils.js [hunk 11]
```diff
@@ -174,4 +175,63 @@ const calculateInterval = (start, end) => {
...
+  Buffer.from(`${
+    Buffer.from(`12${`0${(text.length).toString(16)}`.slice(-2)
+    }`, 'hex')}${text}`
+  )
+
+const validateAliases = async (aliases, identifier, dapi) => {
+  const aliasesWithContestedState = await Promise.all(aliases.map(async (alias) => {
+    const [label, domain] = alias.split('.')
+
+    const normalizedLabel = label.toLowerCase().replace(/[oli]/g, (match) => {
```
Comment (pshenmic) [inline packages/api/src/utils.js:188]: use standarized function to convert normalizedLabel

**Review by pshenmic — CHANGES_REQUESTED** [review]

File: packages/api/package.json (hunk 1 above)
Comment (owl352) [reply packages/api/package.json:20]: ✅

File: packages/api/src/DAPI.js (hunk 2 above)
Comment (owl352) [reply packages/api/src/DAPI.js:29]: ✅

File: packages/api/src/DAPI.js (hunk 3 above)
Comment (owl352) [reply packages/api/src/DAPI.js:30]: ✅

File: packages/api/src/DAPI.js (hunk 5 above)
Comment (owl352) [reply packages/api/src/DAPI.js:41]: already

File: packages/api/src/constants.js (hunk 6 above)
Comment (owl352) [reply packages/api/src/constants.js:8]: ✅

File: packages/api/src/utils.js (hunk 11 above)
Comment (owl352) [reply packages/api/src/utils.js:188]: ✅

File: packages/api/src/controllers/IdentitiesController.js (hunk 9 above)
Comment (owl352) [reply packages/api/src/controllers/IdentitiesController.js:35]: This logic is needed to get valid alias data from dapi, but if we don't specify the domain of the name, we can't request data from dapi, so we request data from indexer and then use it to get `contestedState`

File: packages/api/src/controllers/IdentitiesController.js (hunk 8 above)
Comment (owl352) [reply packages/api/src/controllers/IdentitiesController.js:29]: ✅

File: packages/api/src/DAPI.js (hunk 4 above)
Comment (owl352) [reply packages/api/src/DAPI.js:27]: ✅

File: packages/api/src/controllers/IdentitiesController.js (hunk 7 above)
Comment (owl352) [reply packages/api/src/controllers/IdentitiesController.js:22]: ✅

**Review by pshenmic — CHANGES_REQUESTED** [review]

File: packages/api/src/dao/IdentitiesDAO.js [hunk 12]
```diff
@@ -68,18 +71,43 @@ module.exports = class IdentitiesDAO {
...
 
-    return Identity.fromRow({ ...row })
+    if (!row) {
+      return null
+    }
+
+    const identity = Identity.fromRow(row)
+
+    const aliases = await Promise.all(identity.aliases.map(async alias => {
+      const contested = await getAliasInfo(alias, this.dapi)
```
Comment (pshenmic) [inline packages/api/src/dao/IdentitiesDAO.js:81]: const aliasInfo

File: packages/api/src/dao/IdentitiesDAO.js [hunk 13]
```diff
@@ -142,20 +171,40 @@ module.exports = class IdentitiesDAO {
...
-      owner: row.identity_owner,
-      total_data_contracts: parseInt(row.total_data_contracts),
-      total_documents: parseInt(row.total_documents),
-      total_txs: parseInt(row.total_txs)
-    })), page, limit, totalCount)
+    const resultSet = await Promise.all(rows.map(async row => {
+      const balance = await this.dapi.getIdentityBalance(row.identifier.trim())
+
+      const aliases = await Promise.all((row.aliases ?? []).map(async alias => {
+        const contested = await getAliasInfo(alias, this.dapi)
```
Comment (pshenmic) [inline packages/api/src/dao/IdentitiesDAO.js:184]: const aliasInfo

File: packages/api/src/models/Transaction.js [hunk 14]
```diff
@@ -41,6 +56,16 @@ module.exports = class Transaction {
...
 
-    return new Transaction(tx_hash, index, block_hash, block_height, type, data, timestamp, parseInt(gas_used), status, decodedError ?? error)
+    return new Transaction(
+      tx_hash, index, block_hash,
+      block_height, type, data,
+      timestamp, parseInt(gas_used),
+      status, decodedError ?? error,
+      aliases
+        ? {
+            identifier: owner?.trim(),
```
Comment (pshenmic) [inline packages/api/src/models/Transaction.js]: Use same structure

File: packages/api/src/utils.js [hunk 15]
```diff
@@ -174,4 +176,51 @@ const calculateInterval = (start, end) => {
   }, intervalsInRFC[0])
 }
 
-module.exports = { hash, decodeStateTransition, getKnex, checkTcpConnect, calculateInterval }
+const generateNameIndexBuffer = (name) => {
```
Comment (pshenmic) [inline packages/api/src/utils.js:179]: buildIndexBuffer

**Review by pshenmic — CHANGES_REQUESTED** [review]

**Review by pshenmic — APPROVED** [review]

**Pull request merged**